Adding Chain module to stdlib and tests.

stephentetley opened the pull request:
This PR adds a Chain module to the StdLib - Chain is a List datatype with efficient `snoc` and `append`.

See issue - Library: Add Chain Data Structure? #1543 

The PR isn't quite ready for the mainline, some points for debate are:

Chain uses a custom `ToString` implementation that prints output in the style of `Set#{1, 2, 3}` although there is no corresponding syntax to create chains. Is this a valid? Because a Chain is essentially a binary tree default output would be hard to decipher.

Chain has `fromList`, etc. functions in the module rather than adds `toChain` to the List module. I think this should be changed but there is perhaps a discussion to be had whether "minor" (i.e. less frequently used) modules should add dependencies into the major modules like List and Array.

Ideally Chain's constructors (`Empty`, `One` and `Chain`) should be private to the module (and pattern matching should use `ViewLeft` and `ViewRight`, building should use `empty`, `point` and `fromList`) Is there a way of hiding constructors in Flix at present?

mlutze commented:
I think the `ToString` seems nice; definitely better than showing the tree. I actually wonder if we should make this a more official syntactic sugar where we could interpret `X#{ a, b, c }` as `X.fromList(a :: b :: c :: Nil)`.

As for private constructors, we have a ticket somewhere, but the feature is not implemented.

mlutze commented:
Also, would it be worth it to add benchmarks for this data structure, or should we limit those only to the central ones?

stephentetley commented:
> 
> 
> Also, would it be worth it to add benchmarks for this data structure, or should we limit those only to the central ones?

Benchmarks for building with `append` and `snoc` would certainly be useful. Chain's raison d'etre is that it is supposed to be efficient for building rather than just consing.

magnus-madsen commented:
Great! Looking forward to reviewing it :)


> The PR isn't quite ready for the mainline, some points for debate are:
> 
> Chain uses a custom `ToString` implementation that prints output in the style of `Set#{1, 2, 3}` although there is no corresponding syntax to create chains. Is this a valid? Because a Chain is essentially a binary tree default output would be hard to decipher.

Good question. I don't think we have a principle yet for how `ToString` should work. I remark that Rust has both Debug and ToString. I think in this case they would use ToString for the readable, pretty one, and Debug for something that shows the actual tree structure of the data type.

> Chain has `fromList`, etc. functions in the module rather than adds `toChain` to the List module. I think this should be changed but there is perhaps a discussion to be had whether "minor" (i.e. less frequently used) modules should add dependencies into the major modules like List and Array.

We don't have a solid principle, but we don't really have any `fromX`, but instead lots of `toX`. I think it comes from the observation that if you have a `Foo` and you auto-complete it is nice to see what you can convert it to. (On the other hand, if you know you want a `Foo` it would also seem legitimate to ask from where you can get it... so perhaps the argument is not that strong). Nevertheless I think that for consistency we should try to stay with `toX`. 

Another question is then whether, as you write, `Chain` is sufficient specialized that we should not have `Array.toChain`. But, I think for now we can safely have such functions. If we end up with too many data structures we can consider it then. 
 
> Ideally Chain's constructors (`Empty`, `One` and `Chain`) should be private to the module (and pattern matching should use `ViewLeft` and `ViewRight`, building should use `empty`, `point` and `fromList`) Is there a way of hiding constructors in Flix at present?

Not yet, no :)

mlutze commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,891 @@
+        case SomeRight(Chain[t], t)
+    }
+
+    instance Eq[ViewRight[a]] with Eq[a] {
```
It looks like these instances can be derived from the enum if you want:

```scala
pub enum ViewRight[t] with Eq {
  ...
```

stephentetley replied:
Changed

mlutze commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,891 @@
+    }
+
+    ///
+    /// Add element `x` to the left end of Chain `xs`.
```
`right end`

stephentetley replied:
Changed

stephentetley commented:
I've added a benchmark which hopefully proves Chain's worth for appending and snoc-ing. 

Also in an earlier commit I moved the `fromList` etc. functions into there respective modules and renamed them `toChain`.

~~~

------------------------------ Running Benchmarks ------------------------------

Warmup: ........

Benchmark                                         Iterations          Time (avg)
List.ofArray                                           17400            33 us/op
List.cons                                              27731            61 us/op
List.snoc                                                617         68635 us/op
List.append                                              151        511006 us/op
Chain.ofArray                                          24998            71 us/op
Chain.cons                                             33686           108 us/op
Chain.snoc                                             29260           106 us/op
Chain.append                                            5981           583 us/op

Total time elapsed: 134 seconds. Time budget was: 250 seconds.

Process finished with exit code 0

~~~

magnus-madsen commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def count(f: a -> Bool, xs: Chain[a]): Int32 = countHelper(f, xs, 0)
```
But it is not clear to me if we always see the elements from left to right. If we don't then maybe we should mandate purity.

stephentetley replied:
`count` will always be left to right unless someone changes the implementation. 

As `filter` is already effect polymorphic, does this imply `count` should be? The would also imply the same for List and possibly other data structures.

magnus-madsen commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+    }
+
+    ///
+    /// Deconstruct a Chain from the left end.
```
Could you add a bit more intuition. I assume this means that we are seeing elements from left to right. (And the same for viewRight)?

stephentetley replied:
Fixed for `viewLeft` and `viewRight`

magnus-madsen commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, x1), x2)..., xn)`.
+    ///
+    pub def foldLeft(f: (b, a) -> b & ef, ac: b, xs: Chain[a]): b & ef = match viewLeft(xs) {
```
I think this reveals the hand: When you use viewLeft you must be seeing the elements from left-to-right. 

Nevertheless, some doc further up would be useful! :)

stephentetley replied:
There are more comments on both the exported functions `viewLeft` & `viewRight` and their helper functions.

The comments on the helpers make it explicit that you are traversing left-to-right or right-to-left.

magnus-madsen commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+/// It supports efficient append and "snoc" - appending elements at the tail
+/// of the list.
+///
+/// Note - the constructors `Empty`, `One` and `Chain` should not be used directly.
```
How does Haskell do it? Is the idea that you can export a type without exporting its constructors?

stephentetley replied:
In the modules export list, to export a type with constructors you can do `Chain(..)` to export all the constructors or e.g. `Chain(Empty)` to export the type and only the named constructors.

To export a type and hide all the constructors just list the type name in the module export with no parens / constructor list.

magnus-madsen replied:
OK, thanks. We should think about that.

magnus-madsen commented:
Overall looks great! and thanks for the benchmarks :)

stephentetley commented:
Thanks Magnus, I'm busy this evening but I'll look to work through your comments Thursday/Friday.

stephentetley commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+/*
+ * Copyright 2020
```
Fixed

stephentetley commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+    ///
+    /// Return the singleton chain with element `x`.
+    ///
+    pub def point(x: a): Chain[a] = One(x)
```
Fixed

stephentetley commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+    }
+
+    ///
+    /// Returns the sublist of `xs` without the last element.
```
Fixed

stephentetley commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+    ///
+    /// Returns the length of `xs`.
+    ///
+    pub def length(xs: Chain[a]): Int32 = foldRight((_, ac) -> ac + 1, 0, xs)
```
Fixed, though I didn't change STYLE.md

stephentetley commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+    pub def length(xs: Chain[a]): Int32 = foldRight((_, ac) -> ac + 1, 0, xs)
+
+    ///
+    /// Returns `ys` appended to `xs`.
```
I've changed it to this, though it could still be improved.

~~~

/// Returns a new chain formed by appending `ys` to `xs`.

~~~

stephentetley commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,847 @@
+    ///
+    /// Helper function for `map`.
+    ///
+    def mapHelper(f: a -> b & ef, xs: Chain[a], ac: Chain[b]): Chain[b] & ef = match viewLeft(xs) {
```
A1 - `map` visits the chain left to right.

A2 - the chain has the same structure after map.

`map` could flatten the chain out with the expectation that further processing is most usually left-to-right (the library already has a bit of left-to-right bias e.g. `count`), I've a "natural" inclination against this but I don't have evidence to prove a case. 

I see Chain as a transitional data structure that's lifetime should be as long as it's being constructed but if there is a natural point at which it would be analysed (mapped, counted, ...) and not built further, it should be transformed to a list or array.

magnus-madsen replied:
OK, and thanks for the clarification. I agree with your view. Until evidence emerges to the contrary let us keep it as is.

stephentetley commented:
Hi Magnus, I think this commit clears the issues above.

magnus-madsen commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,865 @@
+    case Chain(Chain[t], Chain[t])
+}
+
+instance ToString[Chain[a]] with ToString[a] {
```
Could I ask you to move the ToString instance after all the other instances?

I appreciate that the other instances are declared in a natural order: SemiGroup, Monoid, Functor into Applicative into Monad. 👍

stephentetley replied:
Fixed.

magnus-madsen commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,865 @@
+    ///
+    /// Returns true if and only if `xs` is the empty chain, i.e. `Empty`.
+    ///
+    pub def isEmpty(xs: Chain[a]): Bool = match xs {
```
I notice that the chains is often called `x` and `xs`. We used to use the same convention for list. 

The most reason idea is that `x` and `xs` inside a function (e.g. pattern match) is fine, but that we now prefer to name the arguments after the name of the "thing". E.g. `l` for `List`, `s` for `Set`, `m` for `Map`. I think here we should always use `c` for `Chain`.

magnus-madsen replied:
(I know that x and xs is standard in the FP community, but I am trying to be a bit more palatable to new programmers and to non-FP programmers, and so I think `c` is better than `xs`).

stephentetley replied:
All should now be changed to `c`.

Where there was an `xs` and `ys` like `eq` they should be `c1` and `c2`. Views should be `v`.

magnus-madsen commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,865 @@
+    ///
+    /// Add element `x` to the left end of Chain `xs`.
+    ///
+    pub def cons(x: a, xs: Chain[a]): Chain[a] = match xs {
```
I think here it is fine to name the element `x` and then name the chains `c` (instead of `xs`).

stephentetley replied:
Fixed

magnus-madsen commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,865 @@
+    def equalsHelper(vx: ViewLeft[a], vy: ViewLeft[a]): Bool with Eq[a] = match (vx, vy) {
+        case (NoneLeft, NoneLeft)                           => true
+        case (SomeLeft(x, xs), SomeLeft(y, ys)) if (x == y) => equalsHelper(viewLeft(xs), viewLeft(ys))
+        case (_, _)                                         => false
```
case _ =>

stephentetley replied:
Fixed.

magnus-madsen commented on main/src/library/Chain.flix:
```diff
@@ -0,0 +1,865 @@
+    ///
+    def scanLeftHelper(f: (b, a) -> b & ef, s: b, xs: Chain[a], acc: Chain[b]): Chain[b] & ef = match viewLeft(xs) {
+        case NoneLeft        => acc
+        case SomeLeft(x, rs) => {let a = f(s, x); scanLeftHelper(f, a, rs, snoc(acc, a))}
```
I notice that you sometimes write it like the above, and other times you break it into multiple lines. 

You don't have to change it, just wanted to remark it :)

stephentetley replied:
I don't have a preference for either (otherwise they would be consistent).

In general it's nice if cases can be one liners, but obviously that isn't always possible.

magnus-madsen reviewed (commented):
I took one more pass over the code. 

It is very clean, but I did still add a few more comments. 

Once those are fixed, I am happy to merge it in :)

magnus-madsen commented:
Merged! Thanks :) :)
